### PR TITLE
provide the ability to modify the prefix of error reporting and ignor…

### PR DESCRIPTION
**Pull request by pweil-**
…e some errors in cmd operations

@ffranz - proposed for https://bugzilla.redhat.com/show_bug.cgi?id=1332432

Provides the ability to not include errors during run which in turn controls Failure or Success message output.  Additionally allows customization of the prefix used during the bulk action so you can choose to report warnings, hints, or anything else in the message.

@openshift/cli-review

### CODE REVIEW

**Comment by pweil-** [comment]
Example output:

```
 $ oadm router
info: password for stats user admin has been set to y1LmlIYswk
--> Creating router router ...
    warning: serviceaccounts "router" already exists
    warning: clusterrolebinding "router-router-role" already exists
    deploymentconfig "router" created
    service "router" created
--> Success

```

**Comment by liggitt** [comment]
this seems like turning human readable messages into an API, which we don't want to do.

File: pkg/cmd/admin/router/router.go [hunk 1]
```diff
@@ -789,7 +790,18 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out, errout io.Write
 		return defaultOutputErr
 	}
 
-	if errs := cfg.Action.WithMessage(fmt.Sprintf("Creating router %s", cfg.Name), "created").Run(list, namespace); len(errs) > 0 {
+	levelFilter := func(e error) string {
+		if match, _ := regexp.MatchString("serviceaccounts|rolebinding.*already exists", e.Error()); match {
```
Comment (liggitt) [inline pkg/cmd/admin/router/router.go]: this seems super fragile

**Comment by pweil-** [comment]
> this seems like turning human readable messages into an API, which we don't want to do.

other suggestions?  I considered just iterating through the list of returned errors in the command itself and parsing them out that way however that doesn't improve any of the output.  It will still show "error:" and "Failure".

File: pkg/cmd/admin/router/router.go (hunk 1 above)
Comment (pweil-) [inline pkg/cmd/admin/router/router.go]: could be anything that you'd consider less fragile, a check against IsAlreadyExists and then looking for SA for example.

**Comment by pweil-** [comment]
Thinking about this a bit more, could you elaborate a bit on your stance here?  This is not touching any part of the message other than the prefix which was hardcoded to "error" and the ability to control reporting "Success" vs "Failure" of a bulk action.  

Most of that is already controllable with the `Bulk.After` anyway, this just allows a little more fine grained control on the default mechanisms (`NewPrintNameOrErrorAfterIndent`).

**Comment by pweil-** [comment]
[test]

File: pkg/cmd/admin/router/router.go (hunk 1 above)
Comment (fabianofranz) [inline pkg/cmd/admin/router/router.go]: Yes, I think you have a well qualified error here, and a `IsAlreadyExists` check from the api would work. Definitely better.

File: pkg/config/cmd/cmd.go [hunk 2]
```diff
@@ -35,13 +35,18 @@ type Mapper interface {
 	InfoForObject(obj runtime.Object, preferredGVKs []unversioned.GroupVersionKind) (*resource.Info, error)
 }
 
+// ErrorFilterFunc provides a way to filter errors during the Bulk.Run.  This may be used in conjunction with
+// BulkAction.WithMessageAndPrefix if you are reporting items as warnings.
+type ErrorFilterFunc func(e error) bool
```
Comment (fabianofranz) [inline pkg/config/cmd/cmd.go]: Not very clear to me what the return value means, does a `true` means "filter out" or "keep it"? Mention it in the doc, or call it `IgnoreErrorFunc`, `BypassErrorFunc`, `FatalErrorFunc`, or something similar.

File: pkg/config/cmd/cmd.go [hunk 3]
```diff
@@ -186,24 +197,31 @@ func (b *BulkAction) DefaultIndent() string {
 	return ""
 }
 
-func (b BulkAction) WithMessage(action, individual string) Runner {
+// PrefixFilter allows customization of the prefix that will be printed for any error that occurs in the BulkAction.
+type PrefixFilter func(e error) string
```
Comment (fabianofranz) [inline pkg/config/cmd/cmd.go]: Maybe `PrefixForError`, `ErrorPrefixMsg`, or something similar?

**Comment by pweil-** [comment]
@fabianofranz updated naming and docs based on feedback. PTAL

File: pkg/cmd/admin/router/router.go [hunk 4]
```diff
@@ -789,12 +789,49 @@ func RunCmdRouter(f *clientcmd.Factory, cmd *cobra.Command, out, errout io.Write
 		return defaultOutputErr
 	}
 
-	if errs := cfg.Action.WithMessage(fmt.Sprintf("Creating router %s", cfg.Name), "created").Run(list, namespace); len(errs) > 0 {
+	levelFilter := func(e error) string {
```
Comment (fabianofranz) [inline pkg/cmd/admin/router/router.go]: `levelPrefixer`?  :)

Comment (pweil-) [inline pkg/cmd/admin/router/router.go]: renamed

**Comment by pweil-** [comment]
https://ci.openshift.redhat.com/jenkins/job/test_pr_origin/10846/

https://github.com/openshift/origin/issues/9355 

re[test]

**Comment by fabianofranz** [comment]
[merge]

**Comment by pweil-** [comment]
[merge]

**Comment by pweil-** [comment]
[merge]

**Pull request merged**